### DISCUSSION TITLE
Can an Okta User be Mocked?

### DISCUSSION THREAD

**Discussion by ej2brown** [cypress-io/cypress#26891] [Questions and Help]
We cannot use a real user for testing [(as shown in the Cypress Okta Authentication docs)](https://docs.cypress.io/guides/end-to-end-testing/okta-authentication#docusaurus_skipToContent_fallback) and are required to mock one. 

Issue: [@okta/okta-auth-js](https://www.npmjs.com/package/@okta/okta-auth-js) cannot be mocked when it's a dependency in another function

### Our normal flow: 
1. OktaAuth is generated on creation and exported
1. We call the API endpoint "Programmatic Login" to receive an auth code to then retrieve the tokens. 
 
### Mock e2e flow: 
1. We intercepted the https calls, i.e. `/api/v1/authn`, using `cy.intercept`. The intercept sets mock tokens and redirects which works beautifully.
2. Unfortunately, the OktaAuth object `cy.stubs` won't seem to work on the OktaAuth object. This means that the object fields cannot be mocked/stub. For example, We call `OktaAuth.isLoginRedirect()` after the redirect. 
```js 
 it("mock okta methods - isLoginRedirect", () => {
    cy.stub(OktaAuth, "isLoginRedirect", (...args) => {
      return true;
    });
    console.assert(OktaAuth.isLoginRedirect() === true);
  });
```
The stubbing of `OktaAuth.isLoginRedirect()` works when called in the cypress test file, but not when it's a dependency in another function - the mock always returns false when the stub is true.

There seem to be related discussions on mocking/stubbing functions within modules (similar to our issue) but I'm not 100% sure if they're related or if this is related to OktaAuth.

### Investigation of Sinon’s 
I also investigated Sinon, which cy.stub uses [stub | Cypress Documentation](https://docs.cypress.io/api/commands/stub#Yields-Icon-namequestion-circle). They outline how to mock a dependency [How to stub a dependency of a module](https://sinonjs.org/how-to/stub-dependency/). 

```js
const sinon = require("sinon");
const OktaAuth = require("components/utils/OktaAuth/OktaAuth");
const OktaAuthLib = require("@okta/okta-auth-js");

describe("login page", () => {
  beforeEach(() => {
    sinon.stub(OktaAuthLib, "isLoginRedirect").returns(true);
    sinon.stub(OktaAuth, "isLoginRedirect").returns(true);
....
```
 

The error I get is isLoginRedirect is non-configurable and non-writable
![image](https://github.com/cypress-io/cypress/assets/55103747/417fc399-445c-4021-90ba-e4c034f20af2)

# My question is can OktaAuth be stubbed in order to mock an Okta user?

---

Specs: 
```json
  {
    "@cypress/webpack-preprocessor": "^5.17.1",
    "@okta/okta-auth-js": "~7.2.0",
    "@okta/okta-react": "~6.7.0",
    "@webpack-cli/serve": "~1.6.0",
    "react": "~18.2.0",
    "cypress": "^12.13.0",
    "webpack": "~5.74.0"
  }
```

**Comment by portyu9** [comment]
Hi @ej2brown 

Yes, you can mock the **authenticated state**, but I would not try to do it by stubbing `@okta/okta-auth-js` from the Cypress spec.

The important distinction is that:

```js
cy.stub(OktaAuth, 'isLoginRedirect')
```

changes the property on the object reference available to the Cypress test. It does not rewrite an Okta import that has already been bundled into the application under test. That explains why your assertion against the object in the spec succeeds while application code continues calling the original implementation.

The `non-configurable and non-writable` error from Sinon is another reason not to use the package export as the test seam.

Current Cypress guidance for Okta uses a different approach: **adapt the application so authenticated state can be supplied while it is running under Cypress**. Cypress's own Real World App does this for programmatic Okta authentication.

Since your `cy.intercept()` setup for `/api/v1/authn` and the token responses is already working, I would keep that boundary mocking and add a small application-level authentication seam. For example, before the normal Okta initialization, have your test build recognize Cypress-provided/mock authentication state and initialize your application's auth state from that instead.

Another option, if your application has its own `AuthService`, is to expose/stub that application-owned service rather than the third-party Okta module.

So the preferred separation is:

- **testing your application with a mocked user:** mock/inject application auth state;
- **testing the Okta integration itself:** use the supported Okta flow (`cy.origin()` or programmatic authentication) with a dedicated test identity.

Trying to replace an imported SDK method from the Cypress spec is brittle because it is not necessarily the same module instance used by the AUT.

Please mark this answer as accepted if it helped, thank you.
